ourjapanlife/findadoc-server
# reworked facilities service

Resolves #222 
Resolves #224 
Resolves #266 
Resolves #267 
Resolves #270

# What changed

- [x] Created a new Results framework. This enables us to better handle errors in the server-side, how to know when to stop if validation fails, giving the user a better error message, and providing consistency for our gql API responses. 🔥 
- [x] Reworked Facilities service 🚧 🏗️ 
    - [x] Rebuilt `getfacilitiesbyid`
        - [x] Added validation ✨ 
        - [x] Added in result design
        - [x] fixed query issue
    - [x] Rebuilt `searchFacilities`/`getFacilities`
        - [x] Added a bunch of validation logic ✨ 
        - [x] Built out new query system. Similar design to submission sec (shoutout @evan-desu for saving me time 🙌 )
        - [x] Added in default values for limit/orderby
        - [x] Added limits to strings and such, validated emails.
        - [x] solved a ton of type issues
    - [x] Rebuilt the `addFacility` function 👷   
        - [x] Reworked all the logic
        - [x] 🦖  Added in new types and migrated to work more with gql types instead of dbTypes. The advantage is now we don't need to clone the object and we can use the user input directly after it's been validated.  
        - [x] Had to rework the `addHealthcareProvider` function as a side-effect/dependency for facility. 🚧 
            - [x] ✅ Improved the type issues
            - [x] ✅ Added in the new result framework
            - [ ] ❌ Skipped over adding validation for a separate PR
            - [x] ✅  You can now save multiple healthcare providers with one facility! 🥳  
- [x] Add pagination capabilities for querying facilities 📖 
- [x] Solved a bunch of various type issues (ids, missing fields, null fields, etc) 😌 
- [x] Created string utils 🧶 
- [x] Updated mock data to work with updated types and new facilities service functions 🏘️ 

# Testing instructions

1. Test out the `getFacilityById` function using the facility query. 
2. Send in proper id.

```
query Facility($facilityId: ID!) {
  facility(id: $facilityId) {
    contact {
      address {
        addressLine1En
        addressLine1Ja
        addressLine2En
      }
      id
      email
      mapsLink
      phone
      website
    }
  }
}

Variables
{
  "facilityId": "youridhere"
}
```



3. Test out the `searchFacilities` function using the facilities query. 
4. Test all the various filter options (see gql schema FacilitiesSearchFilters for options)

Use this query as a starting point

```
query QueryFacilities($filters: FacilitySearchFilters) {
  facilities(filters: $filters) {
    id
    contact {
      address {
        addressLine1En
        addressLine2En
        addressLine1Ja
        cityJa
        cityEn
      }
      email
      id
      mapsLink
      phone
      website
    }
    isDeleted
    nameEn
    nameJa
    healthcareProfessionals {
      acceptedInsurance
      degrees {
        nameJa
        nameEn
      }
      names {
        firstName
        lastName
        locale
        middleName
      }
      specialties {
        names {
          locale
          name
        }
      }
      spokenLanguages {
        nameJa
        nameNative
        nameEn
        iso639_3
      }
    }
    healthcareProfessionalIds
  }
}


Variables
{
  "filters": {
    "limit": 10,
    "offset": 0,
    "nameEn": "Zoo"
  }
}

```
